wordpress-mobile/WordPress-Android
# Category Listing Screen - UI Implementation

Parent #15550

# Description
This PR adds the Implementation of the UI for the Category Listing Screen.  As part of the task, this PR will have the code the following changes in code 

## Additions 

- Adds a new layout XML file `site_settings_categories_row.xml`, Which represents the design for the RecyclerView Row 
- Adds Styles to `styles.xml` for the Layout and TextView in the `site_settings_categories_row` 
- Adds supporting Kotlin classes for RecyclerView  
    - Adapter - `SiteSettingsCategoriesAdapter` 
    - ViewHolder - `SiteSettingsCategoriesViewHolder`
    - Diffcallback  - `SiteSettingsCategoriesDiffCallback`

## Updates  

#### `CategoriesListFragment`
-  Adds function to initialize the Recyclerview with ``SiteSettingsCategoriesAdapter`` and item Decoration 
-  Adds function to show the states according to the Observing Livedata 
-  Adds function to pass to `SiteSettingsCategoriesAdapter` which is invoked on clicking at List Row 

#### `CategoriesListViewModel`
- Adds function, Which will be invoked from `CategoriesListFragment` on clicking at Categories List Row 
- Updates `UiState` class to hold the information about the visibility of views 

#### `site_settings_categories_list_fragment.xml`
- Adds `ProgressBar` to be shown in the loading state


## UI
Screenshots from the Testing Device 

There are three mutually exclusive states for the Category Listing Screen. They are

1. Loading 
2. Content 
3. Error 
    3.1. No Network Error 
    3.2. Generic Error 


| Loading(1)|Content(2)|No Network error(3.1)|Generic Error (3.2)|
| :---        |    :----:   |          ---: |          ---: |
| ![Screenshot_20211116-190107](https://user-images.githubusercontent.com/17463767/142137306-92ba38f7-2bc2-4541-8cb2-f61879aab215.jpg)      | ![Screenshot_20211116-184130](https://user-images.githubusercontent.com/17463767/142137222-e6907595-3856-4544-9c64-728066ba7f8e.jpg)       | ![Screenshot_20211116-190830](https://user-images.githubusercontent.com/17463767/142137401-003503cb-d7a0-43dc-b681-100c2272eb2a.jpg)   | ![Screenshot_20211116-190104](https://user-images.githubusercontent.com/17463767/142137547-b2e7325e-a702-469c-8492-4a51c508ca99.jpg) |


## Yet to be added 
- The navigation on clicking at the Category in the List 
- Show the FAB to create a Category 

These will be taken up in the future in a separate PR

# Testing Instructions

## Prerequisite:
- Go to My Site -> Me  -> App Settings -> Test Feature Configuration
- Enable ManageCategoriesFeatureConfig and restart the app
- Relaunch the app.

## To test:

Test that loading screen is displayed (1) and Content(2) is displayed
- Fresh install the app and logon to the wp.com account 
- Go to my My Site -> Site Settings -> Categories 
- Note that the Screen Displays Loading state first followed by Categories Content 
PS: If no categories are present for the Account, then a default Category - **UnCategorized** exists for an account 

----

These states can be tested by using CategoriesListViewModel tests:

### No Network error(3.1)
#### Categories Screen being accessed for the First Time - 
If the internet is not available, The Site settings page disables all the editing. ie the user won't be able to click the Categories row and navigate to Categories Listing Screen. 
For instance, the Internet is available and the User navigates to the Categories Listing screen and if the Internet is disconnected just before the API call, this Scenario would occur . 

This state can be tested using the Unit test - 
`given no items in db with no internet available, when vm starts, then no connection error is displayed`

PS: Note In this case the categories were able to fetch successfully at least once, the items would be displayed from DB and no network error won't be displayed
This state can be tested using the Unit test - 
`when vm starts, then list of items from db is displayed`

### Generic Error Failed (3.2)

This state can be tested by using`CategoriesListViewModel` tests:
`given no items in db and api error occurs, when vm starts, then error is displayed`

## Regression Notes

1. Potential unintended areas of impact
None


2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests

3. What automated tests I added (or what prevented me from doing so)
None


PR submission checklist:
